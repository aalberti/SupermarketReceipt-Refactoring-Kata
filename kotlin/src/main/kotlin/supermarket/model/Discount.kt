package supermarket.model

open class Discount(open val description: String, open val discountAmount: Double)
data class ProductDiscount(
    val product: Product,
    override val description: String,
    override val discountAmount: Double
) :
    Discount(description, discountAmount)
