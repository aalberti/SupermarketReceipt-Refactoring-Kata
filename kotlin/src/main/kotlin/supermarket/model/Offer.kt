package supermarket.model

abstract class Offer(val product: Product) {
    abstract fun discount(
        catalog: SupermarketCatalog,
        productQuantity: ProductQuantity
    ): Discount?
}

class ThreeForTwoOffer(product: Product) :
    Offer(product) {
    override fun discount(
        catalog: SupermarketCatalog,
        productQuantity: ProductQuantity
    ): Discount? =
        if (productQuantity.quantity > 2) {
            val unitPrice = catalog.getUnitPrice(productQuantity.product)
            val numberOfDiscounts = productQuantity.quantity / 3
            val discountAmount =
                productQuantity.quantity * unitPrice - (numberOfDiscounts * 2.0 * unitPrice + productQuantity.quantity % 3 * unitPrice)
            Discount(productQuantity.product, "3 for 2", discountAmount)
        } else null
}

class TenPercentDiscountOffer(product: Product, private val percentDiscount: Double) :
    Offer(product) {
    override fun discount(
        catalog: SupermarketCatalog,
        productQuantity: ProductQuantity
    ): Discount =
        Discount(
            productQuantity.product, "$percentDiscount% off",
            productQuantity.quantity * catalog.getUnitPrice(productQuantity.product) * percentDiscount / 100.0
        )
}

class TwoForAmountOffer(product: Product, private val amount: Double) :
    Offer(product) {
    override fun discount(
        catalog: SupermarketCatalog,
        productQuantity: ProductQuantity
    ): Discount? =
        if (productQuantity.quantity >= 2) {
            val unitPrice = catalog.getUnitPrice(productQuantity.product)
            val total =
                amount * (productQuantity.quantity.toInt() / 2) + productQuantity.quantity % 2 * unitPrice
            val discountAmount = unitPrice * productQuantity.quantity - total
            Discount(productQuantity.product, "2 for $amount", discountAmount)
        } else null
}

class FiveForAmountOffer(product: Product, private val amount: Double) :
    Offer(product) {
    override fun discount(
        catalog: SupermarketCatalog,
        productQuantity: ProductQuantity
    ): Discount? =
        if (productQuantity.quantity >= 5) {
            val unitPrice = catalog.getUnitPrice(productQuantity.product)
            val numberOfDiscounts = productQuantity.quantity.toInt() / 5
            val discountTotal =
                unitPrice * productQuantity.quantity - (amount * numberOfDiscounts + productQuantity.quantity % 5 * unitPrice)
            Discount(productQuantity.product, "5 for $amount", discountTotal)
        } else null
}