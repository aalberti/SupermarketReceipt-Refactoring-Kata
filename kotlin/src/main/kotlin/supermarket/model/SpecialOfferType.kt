package supermarket.model

enum class SpecialOfferType {
    ThreeForTwo {
        override fun applies(productQuantity: ProductQuantity): Boolean =
            productQuantity.quantity.toInt() > 2
    }, TenPercentDiscount {
        override fun applies(productQuantity: ProductQuantity): Boolean = true
    }, TwoForAmount {
        override fun applies(productQuantity: ProductQuantity): Boolean =
            productQuantity.quantity.toInt() >= 2
    }, FiveForAmount {
        override fun applies(productQuantity: ProductQuantity): Boolean =
            productQuantity.quantity.toInt() >= 5
    };

    abstract fun applies(productQuantity: ProductQuantity): Boolean
}
