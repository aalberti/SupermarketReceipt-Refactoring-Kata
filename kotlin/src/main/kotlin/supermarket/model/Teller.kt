package supermarket.model

import java.util.HashMap

class Teller(private val catalog: SupermarketCatalog) {
    private val offers = HashMap<Product, Offer>()

    fun addSpecialOffer(offer: Offer) {
        this.offers[offer.product] = offer
    }

    fun checksOutArticlesFrom(cart: ShoppingCart): Receipt {
        val receipt = Receipt()
        val productQuantities = cart.items
        for (pq in productQuantities) {
            val p = pq.product
            val quantity = pq.quantity
            val unitPrice = this.catalog.getUnitPrice(p)
            val price = quantity * unitPrice
            receipt.addProduct(p, quantity, unitPrice, price)
        }
        for (item in cart.items) {
            offers[item.product]?.let { offer ->
                offer.discount(catalog, item)?.let { discount -> receipt.addDiscount(discount) }
            }
        }

        return receipt
    }
}
