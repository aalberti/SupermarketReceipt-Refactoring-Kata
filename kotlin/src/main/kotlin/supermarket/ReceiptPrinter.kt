package supermarket

import supermarket.model.*
import java.util.*

class ReceiptPrinter @JvmOverloads constructor(private val columns: Int = 40) {

    fun printReceipt(receipt: Receipt): String {
        val result = StringBuilder()
        for (item in receipt.getItems()) {
            val price = String.format(Locale.UK, "%.2f", item.totalPrice)
            val quantity = presentQuantity(item)
            val name = item.product.name
            val unitPrice = String.format(Locale.UK, "%.2f", item.price)

            val whitespaceSize = columns - name.length - price.length
            var line = name + getWhitespace(whitespaceSize) + price + "\n"

            if (item.quantity != 1.0) {
                line += "  $unitPrice * $quantity\n"
            }
            result.append(line)
        }
        result.append(discountsDescription(receipt.getDiscounts()))
        result.append("\n")
        val pricePresentation = String.format(Locale.UK, "%.2f", receipt.totalPrice as Double)
        val total = "Total: "
        val whitespace = getWhitespace(columns - total.length - pricePresentation.length)
        result.append(total).append(whitespace).append(pricePresentation)
        return result.toString()
    }

    private fun discountsDescription(discounts: List<Discount>) =
        if (discounts.isEmpty()) ""
        else discounts.joinToString("\n") { discountDescription(it) } + "\n"

    private fun discountDescription(discount: Discount): String {
        val productPresentation = if (discount is ProductDiscount) "(${discount.product.name})" else ""
        val pricePresentation = String.format(Locale.UK, "%.2f", discount.discountAmount)
        val description = discount.description
        val whitespace =
            getWhitespace(columns - 1 - productPresentation.length - description.length - pricePresentation.length)
        return "$description$productPresentation$whitespace-$pricePresentation"
    }

    private fun presentQuantity(item: ReceiptItem): String {
        return if (ProductUnit.Each == item.product.unit)
            String.format("%x", item.quantity.toInt())
        else
            String.format(Locale.UK, "%.3f", item.quantity)
    }

    private fun getWhitespace(whitespaceSize: Int): String {
        val whitespace = StringBuilder()
        for (i in 0 until whitespaceSize) {
            whitespace.append(" ")
        }
        return whitespace.toString()
    }
}
