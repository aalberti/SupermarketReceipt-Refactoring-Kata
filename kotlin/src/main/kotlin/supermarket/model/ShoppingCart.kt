package supermarket.model

import supermarket.model.SpecialOfferType.*

class ShoppingCart {

    private val items = ArrayList<ProductQuantity>()
    internal var productQuantities: MutableMap<Product, Double> = HashMap()


    internal fun getItems(): List<ProductQuantity> {
        return ArrayList(items)
    }

    internal fun productQuantities(): Map<Product, Double> {
        return productQuantities
    }


    fun addItemQuantity(product: Product, quantity: Double) {
        items.add(ProductQuantity(product, quantity))
        if (productQuantities.containsKey(product)) {
            productQuantities[product] = productQuantities[product]!! + quantity
        } else {
            productQuantities[product] = quantity
        }
    }

    internal fun handleOffers(receipt: Receipt, offers: Map<Product, Offer>, catalog: SupermarketCatalog) {
        for (product in productQuantities().keys) {
            val quantity = productQuantities[product]!!
            if (offers.containsKey(product)) {
                val offer = offers[product]!!
                val discount: Discount? = when (offer.offerType) {
                    TwoForAmount -> {
                        val quantityAsInt = quantity.toInt()
                        if (quantityAsInt >= 2) {
                            val total =
                                offer.argument * (quantityAsInt / 2) + quantityAsInt % 2 * catalog.getUnitPrice(product)
                            val discountN = catalog.getUnitPrice(product) * quantity - total
                            Discount(product, "2 for " + offer.argument, discountN)
                        } else null
                    }
                    ThreeForTwo -> {
                        val quantityAsInt = quantity.toInt()
                        val numberOfXs = quantityAsInt / 3
                        if (quantityAsInt > 2) {
                            val discountAmount =
                                quantity * catalog.getUnitPrice(product) - (numberOfXs.toDouble() * 2.0 * catalog.getUnitPrice(
                                    product
                                ) + quantityAsInt % 3 * catalog.getUnitPrice(product))
                            Discount(product, "3 for 2", discountAmount)
                        } else null
                    }
                    TenPercentDiscount -> {
                        Discount(product, offer.argument.toString() + "% off",
                                quantity * catalog.getUnitPrice(product) * offer.argument / 100.0)
                    }
                    FiveForAmount -> {
                        val quantityAsInt = quantity.toInt()
                        val numberOfXs = quantityAsInt / 5
                        if (quantityAsInt >= 5) {
                            val discountTotal =
                                catalog.getUnitPrice(product) * quantity - (offer.argument * numberOfXs + quantityAsInt % 5 * catalog.getUnitPrice(
                                    product
                                ))
                            Discount(product, 5.toString() + " for " + offer.argument, discountTotal)
                        } else null
                    }
                }
                if (discount != null)
                    receipt.addDiscount(discount)
            }
        }
    }
}
