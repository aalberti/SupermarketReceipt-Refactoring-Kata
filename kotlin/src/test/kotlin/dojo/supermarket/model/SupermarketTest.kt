package dojo.supermarket.model

import org.approvaltests.Approvals.verify
import org.junit.jupiter.api.Test
import supermarket.ReceiptPrinter
import supermarket.model.*

class SupermarketTest {
    private val toothbrush = Product("toothbrush", ProductUnit.Each)
    private val apples = Product("apples", ProductUnit.Kilo)
    private val teller: Teller by lazy {
        val catalog = FakeCatalog()
        catalog.addProduct(toothbrush, 0.99)
        catalog.addProduct(apples, 1.99)
        Teller(catalog)
    }
    private val printer = ReceiptPrinter()

    @Test
    fun `no discount`() {
        val cart = ShoppingCart()
        cart.addItem(apples, 2.5)

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `add twice`() {
        val cart = ShoppingCart()
        cart.addItem(apples, 1.0)
        cart.addItem(apples, 1.0)

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `10 percent discount`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 1.0)
        teller.addSpecialOffer(TenPercentDiscountOffer(toothbrush, 10.0))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `discount for another product doesn't apply`() {
        val cart = ShoppingCart()
        cart.addItem(apples, 1.0)
        teller.addSpecialOffer(TenPercentDiscountOffer(toothbrush, 10.0))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `3 for 2`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 3.0)
        teller.addSpecialOffer(ThreeForTwoOffer(toothbrush))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `3 for 2 doesn't apply`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 2.0)
        teller.addSpecialOffer(ThreeForTwoOffer(toothbrush))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `2 for amount`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 3.0)
        teller.addSpecialOffer(TwoForAmountOffer(toothbrush, 1.50))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `2 for amount twice`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 4.0)
        teller.addSpecialOffer(TwoForAmountOffer(toothbrush, 1.50))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `2 for amount doesn't apply`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 1.0)
        teller.addSpecialOffer(TwoForAmountOffer(toothbrush, 1.50))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `5 for amount`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 6.0)
        teller.addSpecialOffer(FiveForAmountOffer(toothbrush, 2.99))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `5 for amount twice`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 12.0)
        teller.addSpecialOffer(FiveForAmountOffer(toothbrush, 5.99))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }

    @Test
    fun `5 for amount doesn't apply`() {
        val cart = ShoppingCart()
        cart.addItem(toothbrush, 4.0)
        teller.addSpecialOffer(FiveForAmountOffer(toothbrush, 2.99))

        val receipt = teller.checksOutArticlesFrom(cart)

        verify(printer.printReceipt(receipt))
    }
}
