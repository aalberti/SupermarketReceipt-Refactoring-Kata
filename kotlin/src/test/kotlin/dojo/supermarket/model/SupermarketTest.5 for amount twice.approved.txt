toothbrush                         11.88
  0.99 * c
5 for 5.99(toothbrush)            --2.08

Total:                             13.96